pkg/scc/operator.go

```diff
@@ -0,0 +1,162 @@
+package scc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/rancher/rancher/pkg/scc/consts"
+	"github.com/rancher/rancher/pkg/telemetry"
+
+	"github.com/rancher/rancher/pkg/scc/util/log"
+
+	"github.com/pborman/uuid"
+	k8sv1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/rancher/rancher/pkg/generated/controllers/scc.cattle.io"
+	"github.com/rancher/rancher/pkg/scc/controllers"
+	"github.com/rancher/rancher/pkg/scc/systeminfo"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/wrangler"
+	corev1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/v3/pkg/start"
+)
+
+type sccOperator struct {
+	log                     log.StructuredLogger
+	configMaps              corev1.ConfigMapController
+	sccResourceFactory      *scc.Factory
+	secrets                 corev1.SecretController
+	systemInfoProvider      *systeminfo.InfoProvider
+	systemInfoExporter      *systeminfo.InfoExporter
+	systemRegistrationReady chan struct{}
+}
+
+func setup(wContext *wrangler.Context, logger log.StructuredLogger) (*sccOperator, error) {
+	namespaces := wContext.Core.Namespace()
+	var kubeSystemNS *k8sv1.Namespace
+
+	kubeNsErr := retry.OnError(
+		retry.DefaultRetry,
+		func(err error) bool {
+			return apierrors.IsNotFound(err)
+		},
+		func() error {
+			maybeNs, err := namespaces.Get("kube-system", metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			kubeSystemNS = maybeNs
+			return nil
+		},
+	)
+
+	// TODO(dan) : below we need to be more careful about where we logrus.Fatal. This calls os.Exit(1)
+	if kubeNsErr != nil {
+		// fatal log here, because we need the kube-system ns UID while creating any backup file
+		logger.Fatalf("Error getting namespace kube-system: %v", kubeNsErr)
+	}
+
+	rancherUuid := settings.InstallUUID.Get()
+	if rancherUuid == "" {
+		err := errors.New("no rancher uuid found")
+		logger.Fatalf("Error getting rancher uuid: %v", err)
+		return nil, err
+	}
+
+	sccResources, err := scc.NewFactoryFromConfig(wContext.RESTConfig)
+	if err != nil {
+		logger.Fatalf("Error getting scc resources: %v", err)
+		return nil, err
+	}
+	// Validate that the UUID is in correct format
+	parsedRancherUUID := uuid.Parse(rancherUuid)
+	parsedkubeSystemNSUID := uuid.Parse(string(kubeSystemNS.UID))
+
+	if parsedRancherUUID == nil || parsedkubeSystemNSUID == nil {
+		// TODO (dan) : handle error case as you feel is appropriate
+		return nil, fmt.Errorf("invalid UUID format: rancherUuid=%s, kubeSystemNS.UID=%s", rancherUuid, string(kubeSystemNS.UID))
```

Comment:
I think the only way I can see this happening (short of a catastrophic error) would be if this somehow ran before RMS has created an install-uuid; maybe retry somewhat in the future later automatically?

Reply:
Yeah, this will only fail if Rancher install-uuid is missing or the k8s clusterUID is missing. Either case is likely a very big issue happening.

For the rancher UID, it would have to be a very unique corruption affecting that UUID. The reason being that we can't/won't start the operator until both: a) Ranchers server URL is configured, and b) Rancher's local cluster is reporting `nodes > 0`.

AFAIK, there would be very few situations when both of these (already awaited) conditions pass and the `install-uuid` would be empty/invalid.